src/game/logic/object/object.cpp

```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
+
+        if (power != nullptr) {
+            power->Do_Special_Power_Using_Waypoints(wp, i);
+        }
+    }
 }
 
-bool Object::Is_Salvage_Crate() const
+void Object::Do_Command_Button(const CommandButton *button, CommandSourceType type)
 {
-    return false;
-}
+    if (!Is_Disabled()) {
+        AIUpdateInterface *update = Get_AI_Update_Interface();
 
-void Object::Force_Refresh_Sub_Object_Upgrade_Status() {}
+        if (button != nullptr) {
+            switch (button->Get_Command()) {
+                case GUI_COMMAND_DOZER_CONSTRUCT:
+                case GUI_COMMAND_UNIT_BUILD: {
+                    const ThingTemplate *tmplate = *button->Get_Template();
+                    ProductionUpdateInterface *production = Get_Production_Update_Interface();
 
-void Object::Set_Layer(PathfindLayerEnum layer) {}
+                    if (production == nullptr && tmplate == nullptr) {
+                        goto l1;
+                    }
 
-void Object::Friend_Prepare_For_Map_Boundary_Adjust() {}
+                    production->Queue_Create_Unit(tmplate, production->Request_Unique_Unit_ID());
+                    break;
+                }
+                case GUI_COMMAND_PLAYER_UPGRADE:
+                case GUI_COMMAND_OBJECT_UPGRADE: {
+                    const UpgradeTemplate *upgrade = button->Get_Upgrade_Template();
+                    captainslog_dbgassert(upgrade != nullptr, "Undefined upgrade '%s' in player upgrade command", "UNKNOWN");
+
+                    if (upgrade == nullptr
+                        || upgrade->Get_Type() == UPGRADE_TYPE_OBJECT
+                            && (Has_Upgrade(upgrade) || !Affected_By_Upgrade(upgrade))) {
+                        goto l1;
+                    }
 
-void Object::Friend_Notify_Of_New_Map_Boundary() {}
+                    ProductionUpdateInterface *production = Get_Production_Update_Interface();
 
-void Object::Calc_Natural_Rally_Point(Coord2D *pt) {}
+                    if (production == nullptr) {
+                        goto l1;
+                    }
 
-void Object::Mask_Object(bool mask) {}
+                    production->Queue_Upgrade(upgrade);
+                    break;
+                }
+                case GUI_COMMAND_STOP: {
+                    if (update == nullptr) {
+                        goto l1;
+                    }
 
-bool Object::Is_Using_Airborne_Locomotor() const
-{
-    return false;
-}
+                    update->AI_Idle(type);
+                    break;
+                }
+                case GUI_COMMAND_SELL: {
+                    g_theBuildAssistant->Sell_Object(this);
+                    break;
+                }
+                case GUI_COMMAND_FIRE_WEAPON: {
+                    if (update == nullptr) {
+                        goto l1;
+                    }
 
-void Object::Update_Obj_Values_From_Map_Properties(Dict *properties) {}
+                    if ((button->Get_Options() & 7) != 0 || (button->Get_Options() & 0x20) != 0) {
```

Comment:
Do we have any clue on what these bit fields are?

Reply:
No.